src/replica/replication_app_base.cpp

```diff
@@ -45,6 +45,46 @@ namespace replication {
 
 const std::string replica_init_info::kInitInfo = ".init-info";
 
+DEFINE_TASK_CODE_AIO(LPC_AIO_INFO_WRITE, TASK_PRIORITY_COMMON, THREAD_POOL_DEFAULT)
+
+namespace {
+error_code write_blob_to_file(const std::string &file, const blob &data)
+{
+    std::string tmp_file = file + ".tmp";
+    disk_file *hfile = file::open(tmp_file.c_str(), O_WRONLY | O_CREAT | O_BINARY | O_TRUNC, 0666);
+    ERR_LOG_AND_RETURN_NOT_TRUE(hfile, ERR_FILE_OPERATION_FAILED, "open file {} failed", tmp_file);
+
+    error_code err;
+    size_t sz = 0;
+    task_tracker tracker;
+    aio_task_ptr tsk = file::write(hfile,
+                                   data.data(),
+                                   data.length(),
+                                   0,
+                                   LPC_AIO_INFO_WRITE,
+                                   &tracker,
+                                   [&err, &sz](error_code e, size_t s) {
+                                       err = e;
+                                       sz = s;
+                                   },
+                                   0);
+    dassert_f(tsk, "create file::write task failed");
+    tracker.wait_outstanding_tasks();
+    file::flush(hfile);
+    file::close(hfile);
+    ERR_LOG_AND_RETURN_NOT_OK(err, "write file {} failed", tmp_file);
+    dcheck_eq(data.length(), sz);
+    // TODO(yingchun): need fsync too？
+    ERR_LOG_AND_RETURN_NOT_TRUE(utils::filesystem::rename_path(tmp_file, file),
```

Comment:
Should we remove tmp file if `rename_path` returns error?

Reply:
In fact, I prefer to use assert here, defensive programming will make our code too complex, and the upper layer doesn't deal with every failed cases well actually.
So keep the logic the same as before for this PR, we can do more refactor in the future.